crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1150,33 +1151,50 @@ impl<'db> Type<'db> {
                     && iter::zip(first_elements, second_elements).all(equivalent)
             }
 
-            // TODO: Handle equivalent unions with items in different order
             (Type::Union(first), Type::Union(second)) => {
-                let first_elements = first.elements(db);
-                let second_elements = second.elements(db);
+                if first == second {
+                    return true;
+                }
+
+                let first = first.to_sorted_union(db);
+                let second = second.to_sorted_union(db);
 
-                if first_elements.len() != second_elements.len() {
-                    return false;
+                if first == second {
+                    return true;
                 }
 
-                iter::zip(first_elements, second_elements).all(equivalent)
+                let first_elements = first.elements(db);
+                let second_elements = second.elements(db);
+
+                // TODO: `Unknown ≡ Any`, so `int | Unknown ≡ int | Unknown | Any`
+                // even though they have differently sized element lists.
+                first_elements.len() == second_elements.len()
```

Comment:
Can we move the `len` checks before the `to_sorted_union` (same for intersections).